### PR TITLE
Run adam-pipeline in single machine mode (resolves #210, #329, #371, #325, #326)

### LINKED ISSUE #210: Support running ADAM/GATK pipeline on local storage
* Bypass storage to S3
* Run Spark using `local`

### PR DESCRIPTION
**Pull request by fnothaft**
Depends on #328. I presume that we'll get #328 merged, and then sort out this one.

If anyone would like to make a review pass though, the last three commits are "the business".

This is needed for https://github.com/BD2KGenomics/cgl-docker-lib/issues/139.

### CODE REVIEW

File: src/toil_scripts/tools/spark_tools.py [hunk 1]
```diff
@@ -128,28 +131,58 @@ def call_conductor(master_ip, src, dst, memory=None, override_parameters=None):
...
+              local_dir=None,
+              native_adam_path=None):
     """
     Invokes the ADAM container. Find ADAM at https://github.com/bigdatagenomics/adam.
 
     :param masterIP: The Spark leader IP address.
     :param arguments: Arguments to pass to ADAM.
     :param memory: Gigabytes of memory to provision for Spark driver/worker.
     :param override_parameters: Parameters passed by the user, that override our defaults.
+    :param native_adam_path: Path to ADAM executable. If not provided, Docker is used.
```
Comment (fnothaft) [inline src/toil_scripts/tools/spark_tools.py]: TODO: Document `run_local`, `local_dir`

File: src/toil_scripts/tools/spark_tools.py [hunk 2]
```diff
@@ -128,28 +131,58 @@ def call_conductor(master_ip, src, dst, memory=None, override_parameters=None):
...
+    else:
+        master = ["--master",
+                  ("spark://%s:%s" % (master_ip, SPARK_MASTER_PORT)),
+                  "--conf", ("spark.hadoop.fs.default.name=hdfs://%s:%s" % (master_ip, HDFS_MASTER_PORT)),]
+        work_dir = '.'
+
+    default_params = (master + [
+            # set max result size to unlimited, see #177
+            "--conf", "spark.driver.maxResultSize=0",
+            "--conf", "spark.storage.memoryFraction=0.3",
```
Comment (fnothaft) [inline src/toil_scripts/tools/spark_tools.py]: TODO: document these parameters.

File: src/toil_scripts/adam_pipeline/adam_preprocessing.py [hunk 3]
```diff
@@ -303,8 +352,8 @@ def main():
     parser_run.add_argument('--config', default='adam_preprocessing.config', type=str,
                             help='Path to the (filled in) config file, generated with "generate-config". '
                                  '\nDefault value: "%(default)s"')
-    parser_run.add_argument('--sample', help='The full s3 url of the input SAM or BAM file')
-    parser_run.add_argument('--output-dir', default=None,
+    parser_run.add_argument('--sample', help='The full s3 url/path to the input SAM or BAM file')
```
Comment (jvivian) [inline src/toil_scripts/adam_pipeline/adam_preprocessing.py]: This is a bit confusing — In the RNA-seq pipeline there's a top-level tuple of URL schemes that gets interpolated into strings that mention we use URLs as inputs.

File: src/toil_scripts/lib/files.py [hunk 4]
```diff
@@ -21,18 +21,18 @@ def tarball_files(tar_name, file_paths, output_dir='.', prefix=''):
             f_out.add(file_path, arcname=arcname)
 
 
-def copy_files(file_paths, output_dir):
+def __forall_files(file_paths, output_dir, op):
```
Comment (jvivian) [inline src/toil_scripts/lib/files.py]: This is great, love this change.

File: src/toil_scripts/lib/files.py [hunk 5]
```diff
@@ -49,6 +49,26 @@ def copy_file_job(job, name, file_id, output_dir):
     copy_files([fpath], output_dir)
 
 
+def move_files(file_paths, output_dir):
```
Comment (jvivian) [inline src/toil_scripts/lib/files.py]: This is a semi-dangerous function to use now that we have caching. Might want to put a warning in the docstring?

**Comment by jvivian** [comment]
Made a couple nit comments, but aside from that LG2M.  You have a few magic numbers but already have line comments about adding documentation.

File: src/toil_scripts/lib/files.py (hunk 5 above)
Comment (fnothaft) [inline src/toil_scripts/lib/files.py]: Good catch, will address.

File: src/toil_scripts/lib/files.py (hunk 4 above)
Comment (fnothaft) [inline src/toil_scripts/lib/files.py]: Thanks! I was pretty chuffed with it myself.

File: src/toil_scripts/adam_pipeline/adam_preprocessing.py (hunk 3 above)
Comment (fnothaft) [inline src/toil_scripts/adam_pipeline/adam_preprocessing.py]: I believe we only support S3 URLs and file paths here. Let me clean up the docs, though.

Comment (fnothaft) [inline src/toil_scripts/adam_pipeline/adam_preprocessing.py]: This is a bit different from the other paths, since we do the download using conductor.

**Comment by fnothaft** [comment]
Thanks for the comments @jvivian! Just pushed fd214ba to address.

**Comment by fnothaft** [comment]
@jvivian just updated the last commit to fix the test issue.

BTW, our tests rely on files that are only accessible with UCSC AWS creds (probably with --requestor-pays works too, haven't checked there). I'll open a ticket for us to resolve that.

**Comment by jvivian** [comment]
All tests passed, but it looks like there's an import error in `align_and_call` that's referring to the (now deleted) uberscript dir.

**Comment by fnothaft** [comment]
Fixed and pushed.

**Comment by fnothaft** [comment]
@jvivian hold on merging this 30min, I just realized that I did something breathlessly stupid in my last commit.

**Comment by jvivian** [comment]
@fnothaft — No sweat, just ping me.

**Comment by fnothaft** [comment]
Actually fixed this time. Ping for merge after tests pass, @jvivian.

**Comment by jvivian** [comment]
Thanks @fnothaft !

**Pull request merged**